File: arrays_hashing/valid_anagram/valid_anagram_test.py
import pytest
from valid_anagram import valid_anagram


@pytest.mark.parametrize(
    "s, t, expected",
    [
        ("racecar", "carrace", True),  # Example 1: Standard anagram
        ("jar", "jam", False),  # Example 2: Different characters
        ("rat", "car", False),  # Same length, different letters
        ("anagram", "nagaram", True),  # Classic test case
        ("a", "ab", False),  # Different lengths
        ("", "", True),  # Edge case: Empty strings
        ("a", "a", True),  # Edge case: Single character
        ("aabb", "bbaa", True),  # Multiple occurrences
        ("pcc", "pc", False),  # Subset length mismatch
    ],
)
def test_valid_anagram(s, t, expected):
    """
    Verifies valid_anagram identifies anagrams correctly
    across various string lengths and character distributions.
    """
    assert valid_anagram(s, t) == expected

File: arrays_hashing/valid_anagram/valid_anagram_test.go
package validanagram

import "testing"

func TestValidAnagram(t *testing.T) {
	tests := []struct {
		name     string
		s        string
		t        string
		expected bool
	}{
		{
			name:     "Example 1: racecar and carrace",
			s:        "racecar",
			t:        "carrace",
			expected: true,
		},
		{
			name:     "Example 2: jar and jam",
			s:        "jar",
			t:        "jam",
			expected: false,
		},
		{
			name:     "Different lengths",
			s:        "apple",
			t:        "aple",
			expected: false,
		},
		{
			name:     "Empty strings",
			s:        "",
			t:        "",
			expected: true,
		},
		{
			name:     "Same characters different counts",
			s:        "aabb",
			t:        "abab",
			expected: true,
		},
		{
			name:     "One string is a subset of another",
			s:        "a",
			t:        "ab",
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ValidAnagram(tt.s, tt.t)
			if result != tt.expected {
				t.Errorf("ValidAnagram(%q, %q) = %v; want %v", tt.s, tt.t, result, tt.expected)
			}
		})
	}
}
